src/diffusers/pipelines/pag/pag_utils.py

```diff
@@ -0,0 +1,260 @@
+    def _get_pag_scale(self, t):
+        r"""
+        Get the scale factor for the perturbed attention guidance.
+        """
+
+        if self.do_pag_adaptive_scaling:
+            signal_scale = self.pag_scale - self.pag_adaptive_scale * (1000 - t)
+            if signal_scale < 0:
+                signal_scale = 0
+            return signal_scale
+        else:
+            return self.pag_scale
+
+    def _apply_perturbed_attention_guidance(self, noise_pred, do_classifier_free_guidance, guidance_scale, t):
+        r"""
+        Apply perturbed attention guidance to the noise prediction.
+
+        Args:
+            noise_pred (torch.Tensor): The noise prediction tensor.
+            do_classifier_free_guidance (bool): Whether to apply classifier-free guidance.
+            guidance_scale (float): The scale factor for the guidance term.
+            t (int): The current time step.
+
+        Returns:
+            torch.Tensor: The updated noise prediction tensor after applying perturbed attention guidance.
+        """
+        pag_scale = self._get_pag_scale(t)
+        if do_classifier_free_guidance:
+            noise_pred_uncond, noise_pred_text, noise_pred_perturb = noise_pred.chunk(3)
+            noise_pred = (
+                noise_pred_uncond
+                + guidance_scale * (noise_pred_text - noise_pred_uncond)
+                + pag_scale * (noise_pred_text - noise_pred_perturb)
+            )
+        else:
+            noise_pred_uncond, noise_pred_perturb = noise_pred.chunk(2)
+            noise_pred = noise_pred_uncond + pag_scale * (noise_pred_uncond - noise_pred_perturb)
+        return noise_pred
+
+    def _prepare_perturbed_attention_guidance(self, cond, uncond, do_classifier_free_guidance):
+        """
+        Prepares the perturbed attention guidance for the PAG model.
+
+        Args:
+            cond (torch.Tensor): The conditional input tensor.
+            uncond (torch.Tensor): The unconditional input tensor.
+            do_classifier_free_guidance (bool): Flag indicating whether to perform classifier-free guidance.
+
+        Returns:
+            torch.Tensor: The prepared perturbed attention guidance tensor.
+        """
+
+        cond = torch.cat([cond] * 2, dim=0)
+
+        if do_classifier_free_guidance:
+            cond = torch.cat([uncond, cond], dim=0)
+        return cond
+
+    def set_pag_applied_layers(self, pag_applied_layers):
+        r"""
+        set the the self-attention layers to apply PAG. Raise ValueError if the input is invalid.
+        """
+
+        if not isinstance(pag_applied_layers, list):
+            pag_applied_layers = [pag_applied_layers]
+
+        for pag_layer in pag_applied_layers:
+            self._check_input_pag_applied_layer(pag_layer)
+
+        self.pag_applied_layers = pag_applied_layers
+
+    @property
+    def pag_scale(self):
+        """
+        Get the scale factor for the perturbed attention guidance.
+        """
+        return self._pag_scale
```

Comment:
@sunovivid would you think that applying different PAG scales w.r.t the attention layer index could lead to potentially better results?

Reply:
I think precisely perturbing each layer can be helpful.

However, in the case of replacing self-attention with an identity matrix (one case of Perturbed Self-Attention), it would not be possible to apply a different PAG scale per layer, as the PAG scale is applied to the output (predicted epsilon) of the UNet, not to each perturbation. (We don't have a scale when replacing the self-attention map with an identity matrix.)

If we use another perturbation that has some parameters (for example, adding noise or dropout), I think it will be helpful to use different parameters per layer as you suggested.

If you mean applying perturbation to different layers per timestep, it could be helpful since it increases the size of the possible perturbation space, making it more likely to include beneficial perturbations.

Reply:
I think both could be studied further in a future study. PAG v2 perhaps.

Reply:
Of course, we are now working on an advanced version of PAG. Thank you for your suggestions!

Reply:
Would love to collaborate if there's a chance :-)